File: server/src/core/grid.js
/**
 * グリッド — 盤面の規則。純粋。
 *
 * DOM も Three.js も DB も知らない（docs/design/08-extensibility.md）。
 * 当たり判定・エリア判定・オブジェクト判定はすべてタイル単位で行う
 * （docs/design/10-assets.md §5 の寸法規約と対応する）。
 */

/** @typedef {{ width:number, height:number, blocked:[number,number][] }} GridDef */
/** @typedef {{ id:string, kind:string, x:number, y:number, data?:object }} SpaceObject */
/** @typedef {{ id:string, x:number, y:number, w:number, h:number, kind:string, name:string }} Area */

export class Grid {
  /**
   * @param {GridDef} def
   * @param {{ areas?:Area[], objects?:SpaceObject[] }} [extra]
   */
  constructor(def, extra = {}) {
    this.width = def.width;
    this.height = def.height;
    this.areas = extra.areas ?? [];
    this.objects = extra.objects ?? [];

    this._blocked = new Set();
    // 外周は常に壁
    for (let x = 0; x < this.width; x++) {
      this._blocked.add(key(x, 0));
      this._blocked.add(key(x, this.height - 1));
    }
    for (let y = 0; y < this.height; y++) {
      this._blocked.add(key(0, y));
      this._blocked.add(key(this.width - 1, y));
    }
    for (const [x, y] of def.blocked ?? []) this._blocked.add(key(x, y));

    this._objectAt = new Map();
    for (const o of this.objects) this._objectAt.set(key(o.x, o.y), o);
  }

  /** 範囲外も「通行不可」として扱う */
  isBlocked(x, y) {
    const cx = Math.floor(x), cy = Math.floor(y);
    if (cx < 0 || cy < 0 || cx >= this.width || cy >= this.height) return true;
    return this._blocked.has(key(cx, cy));
  }

  areaAt(x, y) {
    return this.areas.find(a => x >= a.x && x < a.x + a.w && y >= a.y && y < a.y + a.h) ?? null;
  }

  objectAt(x, y) {
    return this._objectAt.get(key(Math.floor(x), Math.floor(y))) ?? null;
  }

  objectById(id) {
    return this.objects.find(o => o.id === id) ?? null;
  }
}

const key = (x, y) => x + ',' + y;

File: unity/Assets/Hidamari/Core/Grid.cs
// グリッド — 盤面の規則。純粋。
//
// ★ server/src/core/grid.js の写し。
//   UnityEngine を **参照しない**。理由は2つある。
//     1. Unity を開かずに `dotnet run --project unity/Tests.Headless` で検証できる
//     2. 論理レイヤーが表現レイヤーを知らない、という分離（04 §4.5）がそのまま形になる
//
// ★ 写しなので、規則がこの世に2つある。ずれると「歩いたら引き戻される」になる。
//   shared/golden/movement.tsv がその見張り。

using System.Collections.Generic;

namespace Hidamari.Core
{
    public sealed class SpaceObject
    {
        public string Id;
        public string Kind;
        public int X;
        public int Y;
        public int Rot;
    }

    public sealed class Area
    {
        public string Id;
        public int X, Y, W, H;
        public string Kind;
        public string Name;
    }

    public sealed class Grid
    {
        public readonly int Width;
        public readonly int Height;
        public readonly List<Area> Areas = new List<Area>();
        public readonly List<SpaceObject> Objects = new List<SpaceObject>();

        readonly HashSet<long> _blocked = new HashSet<long>();
        readonly Dictionary<long, SpaceObject> _objectAt = new Dictionary<long, SpaceObject>();

        public Grid(int width, int height, IEnumerable<(int x, int y)> blocked)
        {
            Width = width;
            Height = height;

            // 外周は常に壁。呼び出し側が blocked に入れ忘れても塞がる（grid.js と同じ）
            for (var x = 0; x < Width; x++)
            {
                _blocked.Add(Key(x, 0));
                _blocked.Add(Key(x, Height - 1));
            }
            for (var y = 0; y < Height; y++)
            {
                _blocked.Add(Key(0, y));
                _blocked.Add(Key(Width - 1, y));
            }
            if (blocked != null)
                foreach (var (x, y) in blocked) _blocked.Add(Key(x, y));
        }

        public void AddObject(SpaceObject o)
        {
            Objects.Add(o);
            _objectAt[Key(o.X, o.Y)] = o;
        }

        /// 範囲外も「通行不可」として扱う
        public bool IsBlocked(double x, double y)
        {
            var cx = (int)System.Math.Floor(x);
            var cy = (int)System.Math.Floor(y);
            if (cx < 0 || cy < 0 || cx >= Width || cy >= Height) return true;
            return _blocked.Contains(Key(cx, cy));
        }

        public Area AreaAt(double x, double y)
        {
            foreach (var a in Areas)
                if (x >= a.X && x < a.X + a.W && y >= a.Y && y < a.Y + a.H) return a;
            return null;
        }

        public SpaceObject ObjectAt(double x, double y)
        {
            var k = Key((int)System.Math.Floor(x), (int)System.Math.Floor(y));
            return _objectAt.TryGetValue(k, out var o) ? o : null;
        }

        public SpaceObject ObjectById(string id)
        {
            foreach (var o in Objects) if (o.Id == id) return o;
            return null;
        }

        // タイル座標を1つの long に畳む。文字列キーより速く、割り当ても出ない
        static long Key(int x, int y) => ((long)x << 32) ^ (uint)y;
    }
}
